Add key to `reproject_to` to preserve original date

wtbarnes opened the pull request:
Fixes #8323 

I'm not completely sure if this is the best way to modify the resulting date. I tried to modify the date on the WCS object itself but for some reason I don't seem to be able to modify any of the date properties in-place. Modifying the `GenericMap.meta` object seemed the most straightforward.

Cadair commented on sunpy/map/mapbase.py:
```diff
@@ -3195,6 +3201,10 @@ def reproject_to(self, target_wcs, *, algorithm='interpolation', return_footprin
+        if preserve_date:
+            # date-avg is the time used as the coordinate frame obstime and
+            # will already be set by the target_wcs
+            outmap.meta['date-obs'] = self.date.isot
```
I think the correct way to do this would be with the `_set_date` method: https://github.com/sunpy/sunpy/blob/228be4348fedfa5e5ecb8c898a18dbf06226e547/sunpy/map/mapbase.py#L1031

Some classes override this to fix the metadata in their way.

wtbarnes commented:
I think there are a few things I need to think more carefully about here, namely:

- `target_wcs` might not come from a `.wcs` Map property. What happens if `"DATE-OBS"` is the only date stored in `target_wcs`?
- Probably should avoid modifying  `.meta` after creation of the new map. One option is to modify the output of `target_wcs.to_header()`

wtbarnes commented on sunpy/map/mapbase.py:
```diff
@@ -3132,6 +3133,11 @@ def reproject_to(self, target_wcs, *, algorithm='interpolation', return_footprin
+            If ``True``, the ``'DATE-OBS'`` of the original map will be preserved in the
+            header of the reprojected map. Note that this will not affect the obstime of
+            the reprojected map as this coordinate frame is defined by ``'DATE-AVG`` which
+            will be the obstime of the target WCS.
```
As noted above, I didn't end up using those methods so the docstring is unchanged.

wtbarnes commented:
> Hum I think this could be a slippery slope would creating a new map class for reprojected data with a new attribute be crazy?

Also not a fan of subclasses for reprojected data, but I'm not sure I understand what the advantage would even be of this. The different properties `date` and `reference_date` already allow for what you're suggesting here.

ayshih commented on sunpy/map/mapbase.py:
```diff
@@ -3132,6 +3133,11 @@ def reproject_to(self, target_wcs, *, algorithm='interpolation', return_footprin
+            If ``True``, ``'DATE-OBS'`` of the original map will be preserved in the
+            header of the reprojected map. Note that this will not affect the obstime of
+            the reprojected map as this coordinate frame is defined by ``'DATE-AVG'`` which
+            will be the obstime of the target WCS.
```
```suggestion
            If ``True``, the observation time of the reprojected map is set to the
            observation time of the original map instead of the observation time defined in the
            target WCS.  This does not affect the reprojection itself, which instead depends on
            the reference times of the two coordinate frames.
            Defaults to ``False``.
```
Your description is too jargon-y, incorrect in some cases, and very easy to misinterpret even in the cases that it is correct

wtbarnes replied:
FWIW, based on previous reviews, I think you and @Cadair have different opinions about the level of "jargon" to be included in this docstring. Fight amongst each other if you wish.

wtbarnes commented:
> We discussed this at the community meeting and decided it needs some examples updating / adding.

I've modified the time-distance plot and active region tracking examples to use this keyword. In the latter, I've added a few sentences of explanation as to why. I did not feel like this needed its own example as the AR tracking is one is the best illustration of why you'd want to use it.

ayshih commented on sunpy/map/mapbase.py:
```diff
@@ -3134,8 +3141,16 @@ def reproject_to(self, target_wcs, *, algorithm='interpolation', return_footprin
+            # Explicitly not using _set_date or _set_reference_date. We do not know whether
+            # target_header has a DATE-AVG key and if it does not, we would inadvertently
+            # overwrite DATE-OBS (that we are supposed to be preserving).
+            target_header['DATE-AVG'] = target_header.get('DATE-AVG', 'DATE-OBS')
```
I agree that the `DATE-OBS` key may not be present; I recall that used to be the case for some SUVI files.  I think it's fine for this code to require `DATE-OBS`, but let's explicitly check for this situation – `preserve_date_obs=True` but no `DATE-AVG` or `DATE-OBS` – and raise a helpful error for the user.

wtbarnes replied:
Done. Feel free to reword the exception if you think it does not properly capture the ambiguity of not having either keyword.

Cadair commented on docs/whatsnew/7.1.rst:
```diff
@@ -70,3 +70,14 @@ Fido.search will now continue on error
+Preserving the observation time of a map when reprojecting
+==========================================================
+
+When calling `~sunpy.map.GenericMap.reproject_to`, it is not possible to preserve the original observation time
```
```suggestion
When calling `~sunpy.map.GenericMap.reproject_to`, it is now possible to preserve the original observation time
```
